auth/middleware.go

```diff
@@ -57,23 +67,31 @@ func (k *kratosMiddleware) validateSession(r *http.Request) (*client.Session, er
 		return &client.Session{Active: &activeSession}, nil
 	}
 
-	if strings.HasPrefix(r.URL.Path, "/upstream_push") {
+	if username, password, ok := r.BasicAuth(); ok {
+		cacheKey := basicAuthCacheKey(username, k.basicAuthSeparator, password)
+
+		if val, found := k.authSessionCache.Get(cacheKey); found {
+			if session, ok := val.(*client.Session); ok {
+				return session, nil
+			}
+			logger.Errorf("unexpected value found in auth cache. It is of type [%T]", val)
+		}
+
 		loginFlow, _, err := k.client.FrontendApi.CreateNativeLoginFlow(r.Context()).Execute()
 		if err != nil {
 			return nil, fmt.Errorf("failed to create native login flow: %w", err)
 		}
 
-		username, password, ok := r.BasicAuth()
-		if !ok {
-			return nil, errors.New("endpoint requires basic auth")
-		}
-
 		updateLoginFlowBody := client.UpdateLoginFlowBody{UpdateLoginFlowWithPasswordMethod: client.NewUpdateLoginFlowWithPasswordMethod(username, "password", password)}
 		login, _, err := k.client.FrontendApi.UpdateLoginFlow(r.Context()).Flow(loginFlow.Id).UpdateLoginFlowBody(updateLoginFlowBody).Execute()
 		if err != nil {
 			return nil, fmt.Errorf("failed to update native login flow: %w", err)
```

Comment:
Even on auth failure _(when invalid credential is provided)_, I wanted to cache the response for a slightly shorter period of time _(1 minute?)_ so a faulty downstream would not be able to bombard the central instance.

 However, the error returned doesn't have a reliable way to know that the error occurred due to invalid credentials.

Reply:
I think we can cache login failures for 1m irrespective of reason

Reply:
okay - I guess that's reasonable.